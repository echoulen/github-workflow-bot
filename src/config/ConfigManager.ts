import {Config} from "@type/Config";

export class ConfigManager {
  public static getConfig(): Config {
    return {
      integration: process.env.integration as any,
      repository: {
        url: process.env.repoUrl,
        token: process.env.token,
      },
      notify: {
        enable: process.env.enableNotify === "true",
        token: process.env.notifyToken,
      },
      lgtm: {
        enable: process.env.enableLGTM === "true",
        count: Number(process.env.lgtmCount),
      },
    };
  }
}
