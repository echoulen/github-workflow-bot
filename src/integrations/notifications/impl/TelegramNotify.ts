import {Notify} from "@notifications/Notify";
import {Notification} from "@type/notification/Notification";
import {List} from "immutable";
import {injectable} from "inversify";
import TelegramBot from "node-telegram-bot-api";
import "reflect-metadata";

const regex = new RegExp("/start");

@injectable()
export class TelegramNotify implements Notify {
  private telegramBot: TelegramBot;
  private chatroomIDs: List<number> = List().asMutable();

  public start(token: string): void {
    this.telegramBot = new TelegramBot(token, {polling: true});
    this.telegramBot.onText(regex, async (msg: TelegramBot.Message) => {
      const chatId = msg.chat.id;
      if (this.chatroomIDs.contains(chatId)) {
        this.telegramBot.sendMessage(chatId, "The notification service is running").catch(console.error);
      } else {
        this.chatroomIDs.push(chatId);
        this.telegramBot.sendMessage(chatId, "started the notification service").catch(console.error);
      }
    });
  }

  public newPullRequest(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }

  public updatedPullRequest(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }

  public revisionNeeded(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }

  public autoMerging(notification: Notification): Promise<void> {
    // TODO
    return undefined;
  }
}
