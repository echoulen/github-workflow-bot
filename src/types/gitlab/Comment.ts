export interface Comment {
  id: number;
  body: string;
  attachment: any;
  author: {
    id: number;
    name: string;
    username: string;
    state: string;
    avatar_url: string;
    web_url: string;
  };
  created_at: string;
  updated_at: string;
  system: boolean;
  noteable_id: number;
  noteable_type: "MergeRequest" | any;
  noteable_iid: number;
  resolvable?: boolean;
}
