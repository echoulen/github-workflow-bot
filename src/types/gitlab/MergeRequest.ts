export interface MergeRequest {
  id: number;
  iid: number;
  project_id: number;
  title: string;
  description: string;
  state: "opened" | "merged" | "closed";
  created_at: string;
  updated_at: string;
  target_branch: string;
  source_branch: string;
  upvotes: number;
  downvotes: number;
  author: {
    id: number;
    name: string;
    username: string;
    state: "active" | any;
    avatar_url: string;
    web_url: string;
  };
  squash: boolean;
  assignee: string;
  changes_count: string;
  source_project_id: number;
  target_project_id: number;
  labels: string[];
  merged_by?: {
    id: number;
    name: string;
    username: string;
    state: "active";
    avatar_url: string;
    web_url: string;
  };
  merged_at?: string;
  closed_by: string;
  closed_at: string;
  latest_build_started_at: string;
  latest_build_finished_at: string;
  first_deployed_to_production_at: string;
  pipeline: {
    id: number;
    sha: string;
    ref: string;
    status: "success" | string,
    web_url: string;
  };
  diff_refs: {
    base_sha: string;
    head_sha: string;
    start_sha: string;
  };
  work_in_progress: false;
  milestone: string;
  merge_when_pipeline_succeeds: boolean;
  merge_status: "can_be_merged" | any;
  sha: string;
  merge_commit_sha: string;
  user_notes_count: number;
  discussion_locked: boolean;
  should_remove_source_branch: boolean;
  force_remove_source_branch: boolean;
  web_url: string;
  time_stats: {
    time_estimate: number;
    total_time_spent: number;
    human_time_estimate: number;
    human_total_time_spent: number;
  };
  subscribed: boolean;
}
