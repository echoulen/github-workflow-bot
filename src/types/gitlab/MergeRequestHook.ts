export interface MergeRequestHook {
  object_kind: "merge_request";
  event_type: "merge_request";
  user: {
    name: string;
    username: string;
    avatar_url: string;
  };
  project: {
    id: number,
    name: string;
    description: string;
    web_url: string;
    avatar_url: string;
    git_ssh_url: string;
    git_http_url: string;
    namespace: string;
    visibility_level: number;
    path_with_namespace: string;
    default_branch: string;
    ci_config_path: string;
    homepage: string;
    url: string;
    ssh_url: string;
    http_url: string;
  };
  object_attributes: {
    id: number;
    target_branch: string;
    source_branch: string;
    source_project_id: number;
    author_id: number;
    assignee_id: number;
    title: string;
    created_at: string;
    updated_at: string;
    milestone_id: number;
    state: "opened" | "closed" | "merged";
    merge_status: "can_be_merged" | "can_not_be_merged",
    target_project_id: number;
    iid: number;
    description: string;
    updated_by_id: number;
    merge_error: any;
    merge_params: {
      force_remove_source_branch: string;
    };
    merge_when_pipeline_succeeds: boolean;
    merge_user_id: number;
    merge_commit_sha: string;
    deleted_at?: any;
    in_progress_merge_commit_sha?: string;
    lock_version?: string;
    time_estimate: number;
    last_edited_at: string;
    last_edited_by_id: number;
    head_pipeline_id: number;
    ref_fetched?: boolean;
    merge_jid?: number;
    source: {
      id: number;
      name: string;
      description: string;
      web_url: string;
      avatar_url: string;
      git_ssh_url: string;
      git_http_url: string;
      namespace: string;
      visibility_level: number;
      path_with_namespace: string;
      default_branch: string;
      ci_config_path: string;
      homepage: string;
      url: string;
      ssh_url: string;
      http_url: string;
    };
    target: {
      id: number;
      name: string;
      description: string;
      web_url: string;
      avatar_url: string;
      git_ssh_url: string;
      git_http_url: string;
      namespace: string;
      visibility_level: number;
      path_with_namespace: string;
      default_branch: string;
      ci_config_path: string;
      homepage: string;
      url: string;
      ssh_url: string;
      http_url: string;
    };
    last_commit: {
      id: string;
      message: string;
      timestamp: string;
      url: string;
      author: {
        name: string;
        email: string;
      };
    };
    work_in_progress: boolean;
    total_time_spent: number;
    human_total_time_spent: number;
    human_time_estimate: number;
    url: string;
    action: "open" | "update";
  };
  labels: Array<{
    id: number;
    title: string;
    color: string;
    project_id: number;
    created_at: string;
    updated_at: string;
    template: boolean;
    description: string;
    type: string;
    group_id: number;
  }>;
  changes?: {
    labels: {
      previous: any[],
      current: any[],
    },
  };
  repository: {
    name: string;
    url: string;
    description: string;
    homepage: string;
  };
}
