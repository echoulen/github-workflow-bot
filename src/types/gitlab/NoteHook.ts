export interface NoteHook {
    object_kind: "note";
    user: {
        name: string;
        username: string;
        avatar_url: string;
    };
    project_id: number;
    project: {
        name: string;
        description: string;
        web_url: string;
        avatar_url: string;
        git_ssh_url: string;
        git_http_url: string;
        namespace: string;
        visibility_level: number;
        path_with_namespace: string;
        default_branch: string;
        ci_config_path: string;
        homepage: string;
        url: string;
        ssh_url: string;
        http_url: string;
    };
    object_attributes: {
        id: number;
        note: string;
        noteable_type: "MergeRequest" | any;
        author_id: number;
        created_at: string;
        updated_at: string;
        project_id: number;
        attachment: any;
        line_code: string;
        commit_id: string;
        noteable_id: number;
        system: boolean;
        st_diff: null;
        updated_by_id: number;
        type: "DiscussionNote" | "DiffNote" | null;
        position: string;
        original_position: string;
        resolved_at: string;
        resolved_by_id: number;
        discussion_id: string;
        change_position: string;
        resolved_by_push: string;
        url: string;
    };
    repository: {
        name: string;
        url: string;
        description: string;
        homepage: string;
    };
    "merge_request": {
        id: number;
        target_branch: string;
        source_branch: string;
        source_project_id: number;
        author_id: number;
        assignee_id: number;
        title: string;
        created_at: string;
        updated_at: string;
        milestone_id: string;
        state: "opened" | "closed",
        merge_status: "can_be_merged" | "cannot_be_merged",
        target_project_id: number;
        iid: number;
        description: string;
        updated_by_id: string;
        merge_error: any;
        merge_params: {
            force_remove_source_branch: string;
        };
        merge_when_pipeline_succeeds: boolean;
        merge_user_id: number;
        merge_commit_sha: string;
        deleted_at: string;
        in_progress_merge_commit_sha: string;
        lock_version: string;
        time_estimate: number;
        last_edited_at: string;
        last_edited_by_id: number;
        head_pipeline_id: number;
        ref_fetched: boolean;
        merge_jid: string;
        source: {
            name: string;
            description: string;
            web_url: string;
            avatar_url: null,
            git_ssh_url: string;
            git_http_url: string;
            namespace: string;
            visibility_level: number;
            path_with_namespace: string;
            default_branch: string;
            ci_config_path: string;
            homepage: string;
            url: string;
            ssh_url: string;
            http_url: string;
        };
        target: {
            name: string;
            description: string;
            web_url: string;
            avatar_url: string;
            git_ssh_url: string;
            git_http_url: string;
            namespace: string;
            visibility_level: number;
            path_with_namespace: string;
            default_branch: string;
            ci_config_path: string;
            homepage: string;
            url: string;
            ssh_url: string;
            http_url: string;
        };
        last_commit: {
            id: string;
            message: string;
            timestamp: string;
            url: string;
            author: {
                name: string;
                email: string;
            };
        };
        work_in_progress: boolean;
        total_time_spent: number;
        human_total_time_spent: string;
        human_time_estimate: string;
    };
}
