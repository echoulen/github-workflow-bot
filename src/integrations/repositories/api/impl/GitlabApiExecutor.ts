import {ConfigManager} from "@config/ConfigManager";
import {ApiExecutor} from "@repositories/api/ApiExecutor";
import axios from "axios";
import {List} from "immutable";
import {injectable} from "inversify";
import "reflect-metadata";

@injectable()
export class GitlabApiExecutor implements ApiExecutor {

  private readonly url = ConfigManager.getConfig().repository.url;
  private readonly headers = {"PRIVATE-TOKEN": ConfigManager.getConfig().repository.token};

  public async getBotUser<User>(): Promise<User> {
    const res = await axios.get(`${this.url}/api/v4/user`, {headers: this.headers});
    return res.data;
  }

  public async getPullRequest<PullRequest>(project: number, pr: number): Promise<PullRequest> {
    const res = await axios.get(`${this.url}/api/v4/projects/${project}/merge_requests/${pr}`, {headers: this.headers});
    return res.data;
  }

  public async addComment<T>(project: number, pr: number, comment: string): Promise<T> {
    const res = await axios.post(
      `${this.url}/api/v4/projects/${project}/merge_requests/${pr}/notes`,
      {body: comment},
      {headers: this.headers},
    );
    return res.data;
  }

  public async getComments<T>(project: number, pr: number): Promise<List<T>> {
    const res = await axios.get(`${this.url}/api/v4/projects/${project}/merge_requests/${pr}/notes`, {headers: this.headers});
    return List<T>(res.data);
  }

  public async setLabel(project: number, pr: number, labels: string): Promise<void> {
    await axios.put(
      `${this.url}/api/v4/projects/${project}/merge_requests/${pr}`,
      {labels},
      {headers: this.headers},
    );
  }

  public async merge(project: number, pr: number): Promise<void> {
    await axios.put(`${this.url}/api/v4/projects/${project}/merge_requests/${pr}/merge`, {}, {headers: this.headers});
  }
}
