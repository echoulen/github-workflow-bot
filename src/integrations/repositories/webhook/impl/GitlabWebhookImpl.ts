import {ConfigManager} from "@config/ConfigManager";
import {BeanTypes} from "@constants/BeanType";
import {Label} from "@constants/Label";
import {Message} from "@constants/Message";
import {ApiExecutor} from "@repositories/api/ApiExecutor";
import {updateLabels} from "@repositories/utils/updateLabels";
import {Webhook} from "@repositories/webhook/Webhook";
import {Comment} from "@type/gitlab/Comment";
import {MergeRequest} from "@type/gitlab/MergeRequest";
import {MergeRequestHook} from "@type/gitlab/MergeRequestHook";
import {NoteHook} from "@type/gitlab/NoteHook";
import {User} from "@type/gitlab/User";
import {List} from "immutable";
import {inject, injectable} from "inversify";
import "reflect-metadata";
import {sprintf} from "sprintf-js";

@injectable()
export class GitlabWebhookImpl implements Webhook {

  @inject(BeanTypes.API_EXECUTOR)
  private apiExecutor: ApiExecutor;
  private config = ConfigManager.getConfig();
  private autoMergeSec = 25;

  public isPullRequest(hook: MergeRequestHook): boolean {
    return hook.object_kind === "merge_request";
  }

  public isComment(hook: NoteHook): boolean {
    return hook.object_kind === "note";
  }

  public async processPullRequest(hook: MergeRequestHook): Promise<void> {
    const projectId = hook.object_attributes.source_project_id;
    const mergeRequestId = hook.object_attributes.iid;
    const triggerBranch = hook.object_attributes.target_branch;
    const labels: List<string> = List(hook.labels).map<string>((it) => it.title);
    const defaultBranch = hook.project.default_branch;
    const bot = await this.apiExecutor.getBotUser<User>();
    const mergeRequest = await this.apiExecutor.getPullRequest<MergeRequest>(projectId, mergeRequestId);

    if (!mergeRequestId || triggerBranch !== defaultBranch) {
      return;
    }

    if (hook.object_attributes.action === "open") {
      const username = hook.user.username;
      await this.apiExecutor.addComment(
        projectId,
        mergeRequestId,
        sprintf(Message.AWAIT_REVIEWER, username),
      );
      const newLabels = updateLabels(labels, [Label.AWAIT_REVIEWER], []);
      await this.apiExecutor.setLabel(projectId, mergeRequestId, newLabels);
    } else if (
      hook.object_attributes.action === "update" &&
      hook.user.username !== bot.username &&
      mergeRequest.author.username === hook.user.username &&
      List(labels).contains(Label.REVISION_NEEDED)) {

      const newLabels = updateLabels(labels, [Label.UPDATED], [Label.REVISION_NEEDED]);
      await this.apiExecutor.setLabel(projectId, mergeRequestId, newLabels);
    }
  }

  public async processComment(hook: NoteHook): Promise<void> {
    const bot = await this.apiExecutor.getBotUser<User>();

    if (hook.object_attributes.noteable_type !== "MergeRequest") {
      return;
    }

    if (hook.user.username === bot.username) {
      return;
    }

    const projectId = hook.merge_request.target_project_id;
    const mergeRequestId = hook.merge_request.iid;
    const mergeRequest = await this.apiExecutor.getPullRequest<MergeRequest>(projectId, mergeRequestId);
    const labels = List<string>(mergeRequest.labels);

    if (hook.merge_request.work_in_progress || hook.merge_request.merge_status !== "can_be_merged") {
      return;
    }

    if (hook.object_attributes.type === "DiscussionNote" || hook.object_attributes.type === "DiffNote") {
      await this.apiExecutor.setLabel(
        projectId,
        mergeRequestId,
        updateLabels(labels, [Label.REVISION_NEEDED], [Label.AWAIT_REVIEWER, Label.UPDATED, Label.AWAIT_AUTO_MERGE, Label.ONE_LGTM]),
      );
    } else if (hook.object_attributes.note.toUpperCase().includes("LGTM")) {

      const comments = await this.apiExecutor.getComments<Comment>(projectId, mergeRequestId);
      const lgtmReviewers = comments.filter((it) => it.body.toUpperCase().includes("LGTM")).map((comment) => comment.author.name).toSet();
      const currentLgtmCount = lgtmReviewers.count();
      const awaitAutoMergeMessage = sprintf(Message.AWAIT_AUTO_MERGE, this.autoMergeSec);
      const triggerMessage = comments.find((it) => it.body === awaitAutoMergeMessage);

      if (currentLgtmCount >= this.config.lgtm.count && !triggerMessage) {
        await this.apiExecutor.addComment(projectId, mergeRequestId, awaitAutoMergeMessage);
        const cleanLabels = updateLabels(
          labels,
          [Label.AWAIT_AUTO_MERGE],
          [Label.AWAIT_REVIEWER, Label.ONE_LGTM, Label.REVISION_NEEDED, Label.UPDATED],
        );
        await this.apiExecutor.setLabel(projectId, mergeRequestId, cleanLabels);

        setTimeout(async () => {
          await this.apiExecutor.merge(projectId, mergeRequestId);
          const finalMergeRequest = await this.apiExecutor.getPullRequest<MergeRequest>(projectId, mergeRequestId);
          const finalLabels = List<string>(finalMergeRequest.labels);
          const finalNewLabels = updateLabels(finalLabels, [], [Label.AWAIT_AUTO_MERGE, Label.ONE_LGTM, Label.UPDATED]);
          await this.apiExecutor.setLabel(projectId, mergeRequestId, finalNewLabels);
        }, this.autoMergeSec * 1000);
      } else {
        const newLabels = updateLabels(labels, [Label.AWAIT_REVIEWER, Label.ONE_LGTM], [Label.REVISION_NEEDED]);
        await this.apiExecutor.setLabel(projectId, mergeRequestId, newLabels);
      }
    }
  }
}
