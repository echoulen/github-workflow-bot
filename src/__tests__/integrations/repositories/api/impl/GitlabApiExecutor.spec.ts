import {ConfigManager} from "@config/ConfigManager";
import {GitlabApiExecutor} from "@repositories/api/impl/GitlabApiExecutor";
import {MockComments} from "@tests/test-helper/mock-data/gitlab/MockComments";
import {MockMergeRequest} from "@tests/test-helper/mock-data/gitlab/MockMergeRequest";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

describe("GitlabApiExecutor spec", () => {
  const apiExecutor = new GitlabApiExecutor();
  const url = ConfigManager.getConfig().repository.url;
  const project = 123;
  const pr = 54088;
  const mock = new MockAdapter(axios);

  beforeAll(() => {
    mock.onGet(`${url}/api/v4/user`).reply(200, {id: 123, name: "test-user"});
    mock.onGet(`${url}/api/v4/projects/${project}/merge_requests/${pr}`).reply(200, MockMergeRequest);
    mock.onPost(`${url}/api/v4/projects/${project}/merge_requests/${pr}/notes`).reply(200, {id: 54088, body: "test comment"});
    mock.onGet(`${url}/api/v4/projects/${project}/merge_requests/${pr}/notes`).reply(200, MockComments);
  });

  it("should getBotUser proper", async () => {
    const user = await apiExecutor.getBotUser();
    expect(user).toMatchSnapshot();
  });

  it("should getPullRequest proper", async () => {
    const pullRequest = await apiExecutor.getPullRequest(project, pr);
    expect(pullRequest).toMatchSnapshot();
  });

  it("should addComment proper", async () => {
    const comment = await apiExecutor.addComment(project, pr, "test comment");
    expect(comment).toMatchSnapshot();
  });

  it("should getComments proper", async () => {
    const comments = await apiExecutor.getComments(project, pr);
    expect(comments.toArray()).toMatchSnapshot();
  });

  afterAll(() => {
    mock.reset();
    mock.restore();
  });
});
