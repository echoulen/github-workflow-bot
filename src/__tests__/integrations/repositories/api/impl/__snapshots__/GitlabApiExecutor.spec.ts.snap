// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitlabApiExecutor spec should addComment proper 1`] = `
Object {
  "body": "test comment",
  "id": 54088,
}
`;

exports[`GitlabApiExecutor spec should getBotUser proper 1`] = `
Object {
  "id": 123,
  "name": "test-user",
}
`;

exports[`GitlabApiExecutor spec should getComments proper 1`] = `
Array [
  Object {
    "attachment": null,
    "author": Object {
      "avatar_url": "test avatar url",
      "id": 1,
      "name": "tester",
      "state": "active",
      "username": "test username",
      "web_url": "http://gitlab/tester",
    },
    "body": "test comment",
    "created_at": "2018-11-14T09:34:08.450Z",
    "id": 1001,
    "noteable_id": 4247,
    "noteable_iid": 1333,
    "noteable_type": "MergeRequest",
    "resolvable": false,
    "system": false,
    "updated_at": "2018-11-14T09:34:08.450Z",
  },
  Object {
    "attachment": null,
    "author": Object {
      "avatar_url": "test avatar url",
      "id": 1,
      "name": "tester",
      "state": "active",
      "username": "test username",
      "web_url": "http://gitlab/tester",
    },
    "body": "test comment",
    "created_at": "2018-11-14T09:34:08.450Z",
    "id": 1002,
    "noteable_id": 4247,
    "noteable_iid": 1333,
    "noteable_type": "MergeRequest",
    "resolvable": false,
    "system": false,
    "updated_at": "2018-11-14T09:34:08.450Z",
  },
]
`;

exports[`GitlabApiExecutor spec should getPullRequest proper 1`] = `
Object {
  "assignee": null,
  "author": Object {
    "avatar_url": "https://www.gravatar.com/avatar/12345",
    "id": 70,
    "name": "tester",
    "state": "active",
    "username": "test user",
    "web_url": "http://gitlab.com/errol.lin",
  },
  "changes_count": "3",
  "closed_at": null,
  "closed_by": null,
  "created_at": "2018-11-14T09:33:18.270Z",
  "description": "test description",
  "diff_refs": Object {
    "base_sha": "dcbd43328e86d3292f0611f44aa80cd2c819afa4",
    "head_sha": "40439ca171f2a28f657f0cc01dd3d08dd101321c",
    "start_sha": "dcbd43328e86d3292f0611f44aa80cd2c819afa4",
  },
  "discussion_locked": null,
  "downvotes": 0,
  "first_deployed_to_production_at": null,
  "force_remove_source_branch": true,
  "id": 54088,
  "iid": 1333,
  "labels": Array [],
  "latest_build_finished_at": "2018-11-14T09:33:22.216Z",
  "latest_build_started_at": "2018-11-14T09:31:46.054Z",
  "merge_commit_sha": "18aae599763bb86710f7fd3b7bee38ee6e4d959e",
  "merge_status": "can_be_merged",
  "merge_when_pipeline_succeeds": false,
  "milestone": null,
  "pipeline": Object {
    "id": 15221,
    "ref": "VWEB-1108",
    "sha": "40439ca171f2a28f657f0cc01dd3d08dd101321c",
    "status": "success",
    "web_url": "http://gitlab.com/LT/V3Desktop/pipelines/15221",
  },
  "project_id": 32,
  "sha": "40439ca171f2a28f657f0cc01dd3d08dd101321c",
  "should_remove_source_branch": null,
  "source_branch": "my-branch",
  "source_project_id": 123,
  "squash": false,
  "state": "opened",
  "subscribed": true,
  "target_branch": "develop",
  "target_project_id": 123,
  "time_stats": Object {
    "human_time_estimate": null,
    "human_total_time_spent": null,
    "time_estimate": 0,
    "total_time_spent": 0,
  },
  "title": "test title",
  "updated_at": "2018-11-14T11:01:57.279Z",
  "upvotes": 0,
  "user_notes_count": 4,
  "web_url": "http://gitlab.com/LT/V3Desktop/merge_requests/1333",
  "work_in_progress": false,
}
`;
